File: Problem Solving/Algorithms/Strings/Super Reduced String.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'superReducedString' function below.
#
# The function is expected to return a STRING.
# The function accepts STRING s as parameter.
#


def superReducedString(s):
    # Write your code here
    stack = []
    for ch in s:
        if not stack:
            stack.append(ch)
        elif ch == stack[-1]:
            stack.pop()
        else:
            stack.append(ch)

    if not stack:
        return "Empty String"
    return "".join(stack)


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = input()

    result = superReducedString(s)

    fptr.write(result + '\n')

    fptr.close()

File: Problem Solving/Algorithms/Strings/Super Reduced String.cpp
#include <bits/stdc++.h>
#include <stack>

using namespace std;

/*
 * Complete the 'superReducedString' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */

string superReducedString(string s) {
    stack<char> myStack;
    for (auto ch: s) {
        if (myStack.empty())
            myStack.push(ch);
        else if (ch == myStack.top())
            myStack.pop();
        else
            myStack.push(ch);
    }

    if (myStack.empty())
        return "Empty String";

    string outString = "";
    while (!myStack.empty()) {
        outString = myStack.top() + outString;
        myStack.pop();
    }

    return outString;
}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string s;
    getline(cin, s);

    string result = superReducedString(s);

    fout << result << "\n";

    fout.close();

    return 0;
}
